dora/core/server/worker/src/main/java/alluxio/worker/netty/PacketReadTaskStateMachine.java

```diff
@@ -0,0 +1,489 @@
+  }
+
+  /**
+   * Completes the read request. When the request is closed, we should clean up any temporary
+   * state it may have accumulated.
+   *
+   * @param context context of the request to complete
+   */
+  public void completeRequest(T context) throws Exception {
+    mPacketReader.completeRequest(context);
+  }
+
+  /**
+   * Cancel this packet reading task.
+   */
+  public void cancel() {
+    fireNext(TriggerEvent.CANCELLED);
+  }
+
+  /**
+   * Notify the state machine that a channel exception occur.
+   * @param e
+   */
+  public void occurChannelException(Error e) {
+    fireNext(mTriggerEventsWithParam.mChannelExceptionEvent, e);
+  }
+
+  /**
+   * Writes an error read response to the channel and closes the channel after that.
+   */
+  private void replyError(AlluxioStatusException e) {
+    mChannel.writeAndFlush(RPCProtoMessage.createResponse(e))
+        .addListener(ChannelFutureListener.CLOSE);
+  }
+
+  /**
+   * Writes a success response.
+   */
+  private void replyEof() {
+    Preconditions.checkState(!mContext.isDoneUnsafe());
+    mContext.setDoneUnsafe(true);
+    mChannel.writeAndFlush(RPCProtoMessage.createOkResponse(null))
+        .addListeners(ChannelFutureListener.CLOSE_ON_FAILURE);
+  }
+
+  /**
+   * Writes a cancel response.
+   */
+  private void replyCancel() {
+    Preconditions.checkState(!mContext.isDoneUnsafe());
+    mContext.setDoneUnsafe(true);
+    mChannel.writeAndFlush(RPCProtoMessage.createCancelResponse())
+        .addListeners(ChannelFutureListener.CLOSE_ON_FAILURE);
+  }
+
+  /**
+   * Returns the appropriate {@link DataBuffer} representing the data to send, depending on the
+   * configurable transfer type.
+   *
+   * @param context context of the request to complete
+   * @param channel the netty channel
+   * @param len     The length, in bytes, of the data to read from the block
+   * @return a {@link DataBuffer} representing the data
+   */
+  protected DataBuffer getDataBuffer(T context, Channel channel, long offset, int len)
+      throws Exception {
+    return mPacketReader.getDataBuffer(context, channel, offset, len);
+  }
+
+  /**
+   * Helper method to allow firing triggers within state handler methods.
+   * If the triggers are fired directly within state handler methods, they will likely make
+   * recursive calls and blow up the stack.
+   *
+   * @param triggerEvent the next trigger event to fire
+   */
+  public void fireNext(TriggerEvent triggerEvent) {
+    mNextTriggerEvent.set(() -> mStateMachine.fire(triggerEvent));
+  }
+
+  /**
+   * Helper method to allow firing triggers within state handler methods.
+   * If the triggers are fired directly within state handler methods, they will likely make
+   * recursive calls and blow up the stack.
+   *
+   * @param triggerEvent the next trigger event to fire
+   * @param arg0         the argument to be used
+   * @param <Arg0T>      the type of the argument to be used
+   */
+  public <Arg0T> void fireNext(
```

Comment:
this should be private. external code must not cause state transitions directly, but rather through methods like `cancel` and `occurChannelException`.